src/Domain/MigrationStep/s150_ProductVariationMigration/VariantGroup/VariantGroupMigrator.php

```diff
@@ -0,0 +1,153 @@
+    private $familyRepository;
+
+    public function __construct(
+        VariantGroupRepository $variantGroupRepository,
+        VariantGroupValidator $variantGroupValidator,
+        FamilyCreator $familyCreator,
+        FamilyRepository $familyRepository,
+        ProductMigrator $productMigrator,
+        MigrationCleaner $variantGroupMigrationCleaner,
+        TableMigrator $tableMigrator
+    ) {
+        $this->variantGroupRepository = $variantGroupRepository;
+        $this->tableMigrator = $tableMigrator;
+        $this->variantGroupValidator = $variantGroupValidator;
+        $this->variantGroupMigrationCleaner = $variantGroupMigrationCleaner;
+        $this->familyCreator = $familyCreator;
+        $this->productMigrator = $productMigrator;
+        $this->familyRepository = $familyRepository;
+    }
+
+    public function migrate(SourcePim $sourcePim, DestinationPim $destinationPim): void
+    {
+        $this->tableMigrator->migrate($sourcePim, $destinationPim, 'pim_catalog_group_attribute');
+        $this->tableMigrator->migrate($sourcePim, $destinationPim, 'pim_catalog_product_template');
+
+        $this->removeInvalidVariantGroups($destinationPim);
+
+        $variantGroupCombinations = $this->retrieveVariantGroupCombinationsToMigrate($destinationPim);
+
+        foreach ($variantGroupCombinations as $variantGroupCombination) {
+            $this->migrateVariantGroupCombination($variantGroupCombination, $destinationPim);
+        }
+
+        $this->variantGroupMigrationCleaner->clean($destinationPim);
+
+        $numberOfRemovedInvalidVariantGroups = $this->variantGroupRepository->retrieveNumberOfRemovedInvalidVariantGroups($destinationPim);
+        if ($numberOfRemovedInvalidVariantGroups > 0) {
+            throw new InvalidVariantGroupException($numberOfRemovedInvalidVariantGroups);
+        }
+    }
+
+    /**
+     * Remove softly the invalid variant-groups from the migration by changing their type to a specific one.
+     */
+    private function removeInvalidVariantGroups(DestinationPim $pim): void
+    {
+        $variantGroups = $this->variantGroupRepository->retrieveVariantGroups($pim);
+
+        foreach ($variantGroups as $variantGroup) {
+            if (!$this->variantGroupValidator->isVariantGroupValid($variantGroup, $pim)) {
+                $this->variantGroupRepository->softlyRemoveVariantGroup($variantGroup->getCode(), $pim);
+            }
+        }
+    }
+
+    private function retrieveVariantGroupCombinationsToMigrate(DestinationPim $pim): \Traversable
+    {
+        $variantGroupCombinations = $this->retrieveVariantGroupCombinations($pim);
+
+        foreach ($variantGroupCombinations as $variantGroupCombination) {
+            if ($this->variantGroupValidator->isVariantGroupCombinationValid($variantGroupCombination, $pim)) {
+                yield $variantGroupCombination;
+            } else {
+                $this->removeVariantGroupCombination($variantGroupCombination, $pim);
```

Comment:
that's weird to remove invalid variant group combination here. 
Would it be not easier to have a service to do it and call it directly ?

This use of multiples private doing a lot of things are a code smell to indicate that there are a lack of services.

Reply:
Indeed. But in this particular context the services will not be usable elsewhere, and they will not evolve. But yes for this method I could extract it in a service to facilitate understanding.